Add Check Support for Java 17 Sealed Classes: ClassFanOutComplexitiy

child of https://github.com/checkstyle/checkstyle/issues/14969

Check documentation: https://checkstyle.org/checks/metrics/classfanoutcomplexity.html

---

```
PS D:\CS\test> cat config.xml                                                   
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">
<module name="Checker">
    <property name="charset" value="UTF-8"/>
    <module name="TreeWalker">
        <module name="ClassFanOutComplexity">
            <property name="max" value="0"/>
        </module>
    </module>
</module>
PS D:\CS\test> cat src/Test.java                                                
public class Test implements I, J {  // violation
}

sealed class C permits D {   // should this be a violation ?
}
final class D extends C {   // violation
}
interface I {
}
interface J {
}
PS D:\CS\test> java  -jar checkstyle-10.17.0-all.jar -c config.xml src/Test.java
Starting audit...
[ERROR] D:\CS\test\src\Test.java:1:1: Class Fan-Out Complexity is 2 (max allowed is 0). [ClassFanOutComplexity]
[ERROR] D:\CS\test\src\Test.java:6:1: Class Fan-Out Complexity is 1 (max allowed is 0). [ClassFanOutComplexity]
Audit done.
Checkstyle ends with 2 errors.
PS D:\CS\test> 
```

---

from the doc: 

> Checks the number of other types a given class/record/interface/enum/annotation relies on. Also, the square of this has been shown to indicate the amount of maintenance required in functional programs (on a file basis) at least.

I don't think we should add 1 for the classes in the `permits` list. The super class `C` permitted certain classes to extend it but it doesn't rely on the permitted subclasses. the opposite is correct

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/modifier/RedundantModifierCheckTest.java
@@ -333,4 +333,25 @@ public class RedundantModifierCheckTest
         verifyWithInlineConfigParser(
                 getNonCompilablePath("InputRedundantModifierRecords.java"), expected);
     }
+
+    @Test
+    public void testSealedClasses() throws Exception {
+        final String[] expected = {
+            "11:4: " + getCheckMessage(MSG_KEY, "final"),
+            "11:10: " + getCheckMessage(MSG_KEY, "public"),
+            "11:17: " + getCheckMessage(MSG_KEY, "static"),
+            "16:4: " + getCheckMessage(MSG_KEY, "abstract"),
+            "16:13: " + getCheckMessage(MSG_KEY, "public"),
+            "20:4: " + getCheckMessage(MSG_KEY, "public"),
+            "20:12: " + getCheckMessage(MSG_KEY, "static"),
+            "24:9: " + getCheckMessage(MSG_KEY, "abstract"),
+            "24:18: " + getCheckMessage(MSG_KEY, "public"),
+            "29:4: " + getCheckMessage(MSG_KEY, "public"),
+            "29:11: " + getCheckMessage(MSG_KEY, "static"),
+            "33:4: " + getCheckMessage(MSG_KEY, "public"),
+            "33:11: " + getCheckMessage(MSG_KEY, "static"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputRedundantModifierSealedClasses.java"), expected);
+    }
 }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierSealedClasses.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/modifier/redundantmodifier/InputRedundantModifierSealedClasses.java
@@ -1,0 +1,56 @@
+/*
+RedundantModifier
+
+
+*/
+
+//non-compiled with javac: Compilable with Java21
+package com.puppycrawl.tools.checkstyle.checks.modifier.redundantmodifier;
+
+public sealed interface InputRedundantModifierSealedClasses permits A {
+   final public static String s = null;
+   // 3 violations above:
+   //                  'Redundant 'final' modifier'
+   //                  'Redundant 'public' modifier'
+   //                  'Redundant 'static' modifier'
+   abstract public void m();
+   // 2 violations above:
+   //                  'Redundant 'abstract' modifier'
+   //                  'Redundant 'public' modifier'
+   public  static  sealed interface J permits A {
+   // 2 violations above:
+   //                  'Redundant 'public' modifier'
+   //                  'Redundant 'static' modifier'
+        abstract public void m();
+        // 2 violations above:
+        //                  'Redundant 'abstract' modifier'
+        //                  'Redundant 'public' modifier'
+   }
+   public static sealed class B permits D { }
+   // 2 violations above:
+   //                  'Redundant 'public' modifier'
+   //                  'Redundant 'static' modifier'
+   public static non-sealed class D extends B { }
+   // 2 violations above:
+   //                  'Redundant 'public' modifier'
+   //                  'Redundant 'static' modifier'
+}
+
+interface InputRedundantModifierSealedClassesCorrected {
+    String s = null;
+    void m();
+    sealed interface J permits A {
+        void m();
+    }
+    sealed class B permits A, D { }
+    non-sealed class D extends B { }
+}
+
+final class A extends InputRedundantModifierSealedClassesCorrected.B
+        implements InputRedundantModifierSealedClasses.J,
+        InputRedundantModifierSealedClassesCorrected.J,
+        InputRedundantModifierSealedClasses {
+
+    @Override
+    public void m() {}
+}

